cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -454,6 +456,172 @@ Status HashJoinSchema::CollectFilterColumns(std::vector<FieldRef>& left_filter,
   return Status::OK();
 }
 
+class HashJoinNode;
+
+struct BloomFilterPushdownContext {
+  using BuildFinishedCallback = std::function<Status(size_t, AccumulationQueue)>;
+  using FiltersReceivedCallback = std::function<Status()>;
+  using FilterFinishedCallback = std::function<Status(size_t, AccumulationQueue)>;
+  void Init(HashJoinNode* owner, size_t num_threads, TaskScheduler* scheduler,
+            FiltersReceivedCallback on_bloom_filters_received, bool disable_bloom_filter,
+            bool use_sync_execution);
+
+  Status StartProducing();
+
+  void ExpectBloomFilter() { eval_.num_expected_bloom_filters_ += 1; }
+
+  Status BuildBloomFilter(size_t thread_index, AccumulationQueue batches,
+                          BuildFinishedCallback on_finished);
+
+  Status PushBloomFilter();
+
+  Status ReceiveBloomFilter(std::unique_ptr<BlockedBloomFilter> filter,
+                            std::vector<int> column_map) {
+    bool proceed;
+    {
+      std::lock_guard<std::mutex> guard(eval_.receive_mutex_);
+      eval_.received_filters_.emplace_back(std::move(filter));
+      eval_.received_maps_.emplace_back(std::move(column_map));
+      proceed = eval_.num_expected_bloom_filters_ == eval_.received_filters_.size();
+
+      ARROW_DCHECK(eval_.received_filters_.size() == eval_.received_maps_.size());
+      ARROW_DCHECK(eval_.received_filters_.size() <= eval_.num_expected_bloom_filters_);
+    }
+    if (proceed) {
+      return push_.all_received_callback_();
+    }
+    return Status::OK();
+  }
+
+  Status FilterBatches(size_t thread_index, AccumulationQueue batches,
+                       FilterFinishedCallback on_finished) {
+    eval_.batches_ = std::move(batches);
+    eval_.on_finished_ = std::move(on_finished);
+
+    if (eval_.num_expected_bloom_filters_ == 0)
```

Comment:
Why not check for `disable_bloom_filter_` here too?  Then it seems like you wouldn't have to check in `FilterSingleBatch`

Reply:
Actually I think `FilterSingleBatch` should be checking `num_expected_bloom_filters_` not `disable_bloom_filter_`. Disabling bloom filter disables building it, not receiving it.